File: patterns/_tests/StructuralPatterns/Composite/ruby/integration.test.rb
# ============================================================
#  Integration Tests — Composite: Airline Network
#
#  What we test here:
#    · Full three-level tree builds and resolves correctly end-to-end
#    · Root get_capacity equals the true sum of every airport
#    · Mutations to a leaf propagate up to the root in one call
#    · Removing a region removes all its airports from root totals
#    · Adding a new region doesn't break existing queries
#    · print_tree renders the full tree with correct structure
#
#  Run: rspec integration.test.rb
# ============================================================

require 'rspec'
require 'stringio'
require_relative '../../../../StructuralPatterns/Composite/ruby/ruby'

def capture_output(&block)
  old = $stdout
  $stdout = StringIO.new
  block.call
  $stdout.string.lines.map(&:chomp)
ensure
  $stdout = old
end

def build_network
  lax = Airport.new('LAX', 4)
  lax.add_route('JFK'); lax.add_route('ORD'); lax.add_route('LHR')
  sfo = Airport.new('SFO', 3)
  sfo.add_route('SEA'); sfo.add_route('DEN')
  west = RegionalNetwork.new('West Region')
  west.add(lax); west.add(sfo)

  jfk = Airport.new('JFK', 5)
  jfk.add_route('LAX'); jfk.add_route('LHR'); jfk.add_route('CDG'); jfk.add_route('NRT')
  ord = Airport.new('ORD', 3)
  ord.add_route('LAX'); ord.add_route('JFK')
  east = RegionalNetwork.new('East Region')
  east.add(jfk); east.add(ord)

  network = AirportNetwork.new('Global Network')
  network.add(west); network.add(east)
  [network, west, east, lax, jfk]
end

TOTAL_CAPACITY = 4 + 3 + 5 + 3  # 15
TOTAL_ROUTES   = 3 + 2 + 4 + 2  # 11

# ── End-to-end totals ─────────────────────────────────────────
RSpec.describe 'Full network — totals' do
  let(:tree) { build_network }

  it 'root get_capacity equals sum of every airport' do
    expect(tree[0].get_capacity).to eq(TOTAL_CAPACITY)
  end

  it 'root get_route_count equals sum of every route' do
    expect(tree[0].get_route_count).to eq(TOTAL_ROUTES)
  end

  it 'region capacity is independent of siblings' do
    expect(tree[1].get_capacity).to eq(4 + 3)
  end
end

# ── Mutations propagate up ────────────────────────────────────
RSpec.describe 'Full network — mutations propagate to root' do
  it 'adding an airport increases root capacity' do
    network, west, * = build_network
    before = network.get_capacity
    west.add(Airport.new('SEA', 2))
    expect(network.get_capacity).to eq(before + 2)
  end

  it 'removing an airport decreases root capacity' do
    network, west, _, lax, * = build_network
    before = network.get_capacity
    west.remove(lax)
    expect(network.get_capacity).to eq(before - lax.get_capacity)
  end

  it 'removing a region removes all its airports from totals' do
    network, west, * = build_network
    west_cap    = west.get_capacity
    west_routes = west.get_route_count
    cap_before    = network.get_capacity
    routes_before = network.get_route_count
    network.remove(west)
    expect(network.get_capacity).to    eq(cap_before    - west_cap)
    expect(network.get_route_count).to eq(routes_before - west_routes)
  end
end

# ── Adding a new region ────────────────────────────────────────
RSpec.describe 'Full network — adding a new region' do
  it 'new region resolves from root correctly' do
    network, * = build_network
    before  = network.get_capacity
    central = RegionalNetwork.new('Central Region')
    central.add(Airport.new('DFW', 4))
    central.add(Airport.new('DEN', 3))
    network.add(central)
    expect(network.get_capacity).to eq(before + 4 + 3)
  end
end

# ── print_tree output ─────────────────────────────────────────
RSpec.describe 'Full network — print_tree output' do
  let(:lines) do
    network, * = build_network
    capture_output { network.print_tree }
  end

  it 'root network appears first' do
    expect(lines.first).to include('Global Network')
  end

  it 'all regions appear in output' do
    expect(lines).to include(a_string_including('West Region'))
    expect(lines).to include(a_string_including('East Region'))
  end

  it 'all airports appear in output' do
    %w[LAX SFO JFK ORD].each do |code|
      expect(lines).to include(a_string_including(code))
    end
  end

  it 'airports are indented more than their parent region' do
    west_line = lines.find { |l| l.include?('West Region') }
    lax_line  = lines.find { |l| l.include?('LAX') }
    west_indent = west_line.length - west_line.lstrip.length
    lax_indent  = lax_line.length  - lax_line.lstrip.length
    expect(lax_indent).to be > west_indent
  end
end

File: patterns/_tests/StructuralPatterns/Composite/typescript/integration.test.ts
// ============================================================
//  Integration Tests — Composite: Airline Network
//
//  What we test here:
//    · Full three-level tree builds and resolves correctly end-to-end
//    · Root getCapacity() equals the true sum of every airport
//    · Adding a new region does not break existing queries
//    · Mutations to a leaf propagate up to the root in one call
//    · Removing a region removes all its airports from root totals
//    · print() renders the full tree with correct structure
//
//  Run: npx mocha --require ts-node/register integration.test.ts
// ============================================================

import { describe, it, before } from 'mocha';
import assert from 'assert';
import { captureOutput } from '../../../_helpers/typescript/helpers';
import {
  Airport, RegionalNetwork, AirportNetwork,
} from '../../../../StructuralPatterns/Composite/typescript/typescript';

// ── Tree fixture ──────────────────────────────────────────────
//
//  Global Network
//    West Region
//      LAX  cap=4  routes: JFK, ORD, LHR
//      SFO  cap=3  routes: SEA, DEN
//    East Region
//      JFK  cap=5  routes: LAX, LHR, CDG, NRT
//      ORD  cap=3  routes: LAX, JFK
//
//  Total capacity:   4+3+5+3 = 15
//  Total routes:     3+2+4+2 = 11

function buildNetwork(): { network: AirportNetwork; west: RegionalNetwork; east: RegionalNetwork; lax: Airport; jfk: Airport } {
  const lax = new Airport('LAX', 4);
  lax.addRoute('JFK'); lax.addRoute('ORD'); lax.addRoute('LHR');

  const sfo = new Airport('SFO', 3);
  sfo.addRoute('SEA'); sfo.addRoute('DEN');

  const west = new RegionalNetwork('West Region');
  west.add(lax);
  west.add(sfo);

  const jfk = new Airport('JFK', 5);
  jfk.addRoute('LAX'); jfk.addRoute('LHR'); jfk.addRoute('CDG'); jfk.addRoute('NRT');

  const ord = new Airport('ORD', 3);
  ord.addRoute('LAX'); ord.addRoute('JFK');

  const east = new RegionalNetwork('East Region');
  east.add(jfk);
  east.add(ord);

  const network = new AirportNetwork('Global Network');
  network.add(west);
  network.add(east);

  return { network, west, east, lax, jfk };
}

const TOTAL_CAPACITY = 4 + 3 + 5 + 3; // 15
const TOTAL_ROUTES   = 3 + 2 + 4 + 2; // 11

// ── End-to-end totals ─────────────────────────────────────────
describe('Full network — totals', () => {
  let network: AirportNetwork;
  let west:    RegionalNetwork;

  before(() => ({ network, west } = buildNetwork()));

  it('root getCapacity() equals the sum of every airport', () => {
    assert.strictEqual(network.getCapacity(), TOTAL_CAPACITY);
  });

  it('root getRouteCount() equals the sum of every route', () => {
    assert.strictEqual(network.getRouteCount(), TOTAL_ROUTES);
  });

  it('region getCapacity() is independent of siblings', () => {
    assert.strictEqual(west.getCapacity(), 4 + 3);
  });
});

// ── Mutations propagate up ────────────────────────────────────
describe('Full network — mutations propagate to root', () => {
  it('adding an airport to a region increases root getCapacity()', () => {
    const { network, west } = buildNetwork();
    const before = network.getCapacity();
    west.add(new Airport('SEA', 2));
    assert.strictEqual(network.getCapacity(), before + 2);
  });

  it('removing an airport from a region decreases root getCapacity()', () => {
    const { network, west, lax } = buildNetwork();
    const before = network.getCapacity();
    west.remove(lax);
    assert.strictEqual(network.getCapacity(), before - lax.getCapacity());
  });

  it('removing a whole region removes all its airports from root totals', () => {
    const { network, west } = buildNetwork();
    const westCap    = west.getCapacity();
    const westRoutes = west.getRouteCount();
    const capBefore  = network.getCapacity();
    const routesBefore = network.getRouteCount();

    network.remove(west);

    assert.strictEqual(network.getCapacity(),    capBefore    - westCap);
    assert.strictEqual(network.getRouteCount(),  routesBefore - westRoutes);
  });
});

// ── Adding a new level doesn't break existing queries ─────────
describe('Full network — adding a new region', () => {
  it('inserting a new region still resolves from root correctly', () => {
    const { network } = buildNetwork();
    const before = network.getCapacity();

    const central = new RegionalNetwork('Central Region');
    central.add(new Airport('DFW', 4));
    central.add(new Airport('DEN', 3));
    network.add(central);

    assert.strictEqual(network.getCapacity(), before + 4 + 3);
  });
});

// ── print() renders the full tree ────────────────────────────
describe('Full network — print() output', () => {
  let lines: string[];

  before(() => {
    const { network } = buildNetwork();
    lines = captureOutput(() => network.print());
  });

  it('root network appears first', () => {
    assert.ok(lines[0].includes('Global Network'));
  });

  it('all regions appear in output', () => {
    assert.ok(lines.some(l => l.includes('West Region')));
    assert.ok(lines.some(l => l.includes('East Region')));
  });

  it('all airports appear in output', () => {
    ['LAX', 'SFO', 'JFK', 'ORD'].forEach(code => {
      assert.ok(lines.some(l => l.includes(code)), `expected ${code} in output`);
    });
  });

  it('airports are indented more than their parent region', () => {
    const westLine = lines.find(l => l.includes('West Region'));
    const laxLine  = lines.find(l => l.includes('LAX'));
    assert.ok(westLine && laxLine);

    const westIndent = westLine.length - westLine.trimStart().length;
    const laxIndent  = laxLine.length  - laxLine.trimStart().length;
    assert.ok(laxIndent > westIndent);
  });
});
